### PR TITLE
config_stripper no longer stores full file in RAM.

### PR DESCRIPTION
**Pull request by ralimi**
Previously, config_stripper would store the full
TAR file and gzipped TAR file in memory to compute
the SHA256 hash.

After this change, the stripped TAR file is stored
in a temporary location on disk and then renamed
into its final location.  Furthermore, the SHA256
hash of the gzipped TAR file is computed
incrementally as the gzipping progresses.

### CODE REVIEW

File: docker/util/config_stripper.py [hunk 1]
```diff
@@ -114,44 +117,94 @@ def strip_layer(path):
...
+    # Keep track of sha hash for both the compressed and uncompressed tar
+    uncompressed_sha = hashlib.sha256()
+    compressed_sha = hashlib.sha256()
 
-    # Compress buf to gz
+    # Start a gzip process that we'll use to compress tar output.
     # Shelling out to bash gzip is noticeably faster than using python's gzip
+    #
+    # This function takes special care to never store the full tar file or
+    # gzip'd tar file in memory.  Images can be quite large.
```
Comment (alex1545) [inline docker/util/config_stripper.py]: Remove extra space

File: docker/util/config_stripper.py [hunk 2]
```diff
@@ -114,44 +117,94 @@ def strip_layer(path):
...
-    gz = gzip_process.communicate(input=tar)[0]
 
-    # Calculate sha of layer
-    sha = hashlib.sha256(gz).hexdigest()
-    new_name = 'sha256:%s' % sha
-    with open(os.path.join(original_dir, new_name), 'wb') as out:
-      out.write(gz)
+    # Read the gzip'd output and accumulate the sha hash, and save the
+    # compressed copy under the new name.
+    gzip_stdout_exc = []
```
[the remark this answers is not in the document]
Comment (ralimi) [reply docker/util/config_stripper.py]: Without the list, an assignment to the variable will not be visible from outside the do_gzip_stdout function.

File: docker/util/config_stripper.py [hunk 3]
```diff
@@ -114,44 +117,94 @@ def strip_layer(path):
...
+        gzip_process.stdin.close() # Causes gzip to terminate.
+        gzip_stdout.join() # Terminates after gzip closes stdout.
+        gzip_stderr.join() # Terminates after gzip closes stderr.
+        gzip_process.wait() # gzip terminated by now.
+
+    # Check if any of our threads or processes failed.
+    if gzip_stdout_exc:
+        raise gzip_stdout_exc[0]
+    if gzip_process.returncode != 0:
+        raise RuntimeError('gzip process failed: %s' % gzip_stderr_buf.getvalue())
```
[the remark this answers is not in the document]
Comment (ralimi) [reply docker/util/config_stripper.py]: Certainly. Done.

File: docker/util/config_stripper.py (hunk 1 above)
Comment (ralimi) [reply docker/util/config_stripper.py]: Done.

**Comment by ralimi** [comment]
Thank you!

**Pull request merged**